Fix lambda lowering for cast to generic ReadOnlySpan<T>

gafter opened the pull request:
Fixes #31685

The problem to be solved is to generate good code for
``` c#
ReadOnlySpan<byte> s = new byte[] { 1, 2, 3 };
```

There is an implicit conversion operator declared in `ReadOnlySpan<T>` that converts from `T[]` to `ReadOnlySpan<T>`.  However, in the particular case above we would like to elide creation of the array and create a read-only span that wraps the underlying data in the assembly.  There are primitives for doing this.

This was previously done in the compiler by eliding the lowering of the conversion to a bound call.  Lowering would instead preserve the user-defined conversion in the tree.  Then, in emit we would recognize that particular conversion and either perform the optimization if possible, or simply produce a call to the conversion method.  However, that had an unfortunate side-effect (bug), which this PR addresses.

Normally, the lowering phase would translate an invocation of a user-defined conversion operator into a `BoundCall`.  Subsequent phases of the compiler that need to rewrite symbols (such as lambda lowering and iterator rewriting, when the enclosing method is generic) would inherit the proper handling of rewriting symbols from the bound tree rewriter that is generated from `BoundNodes.xml`.  Unfortunately, the symbol appearing inside the conversion is not rewritten by that generated code, and there is no simple way to make it do so.  The bug report #31685 describes symptoms of this underlying issue.

There are a number of approaches one could take to address this issue:
1. Attempt to rehabilitate the original approach by extending the generated rewriter to handle conversions that may contain symbols.  That would require extending a few internal APIs and the bound tree rewriter.
2. Have the emit phase recognize a call to this particular conversion operator, and perform the optimization on that basis.
3. Add a new `BoundNode` that represents this particular optimizable pattern (which we would lower to), and let the generated bound node rewriter handle the contained symbol in subsequent phases where needed.

The approach here takes the third approach.  We back out the implementation of the optimization and add a new implementation based on a newly added bound node.

333fred commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -646,7 +646,15 @@
+
+  <!-- A special node for emitting the implicit conversion from an array (of T) to System.ReadOnlySpan<T> -->
+  <Node Name="BoundReadOnlySpanFromArray" Base="BoundExpression">
+    <!-- The type is required an instance of the well-known type System.ReadOnlySpan<T> -->
```
>required an [](start = 21, length = 11)

"to be"?

333fred commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:
```diff
@@ -1025,16 +1025,18 @@ private static BoundExpression NullableAlwaysHasValue(BoundExpression expression
-                doNotLowerToCall
-                ? BoundConversion.Synthesized(syntax, rewrittenOperand, conversion, false, explicitCastInCode: true, conversionGroupOpt: null, default(ConstantValue), rewrittenType)
-                : (BoundExpression)BoundCall.Synthesized(syntax, null, conversion.Method, rewrittenOperand);
+            // do not rewrite user defined conversion in expression trees or
```
>expression trees or [](start = 57, length = 19)

Cut off sentence?

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadOnlySpanConstructionTest.cs:
```diff
@@ -524,5 +524,123 @@ .maxstack  3
+    }
+}
+", WithNonNullTypesTrue(TestOptions.ReleaseExe));
+            var cv = CompileAndVerify(comp, expectedOutput: "", verify: Verification.Passes);
```
>verify: Verification.Passes [](start = 64, length = 27)

nit: could be removed

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -646,7 +646,15 @@
+  <Node Name="BoundReadOnlySpanFromArray" Base="BoundExpression">
+    <!-- The type is required an instance of the well-known type System.ReadOnlySpan<T> -->
+    <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
+    <Field Name="Operand" Type="BoundExpression"/>
```
>BoundExpression [](start = 32, length = 15)

Can this be a more specific type? (`BoundArrayCreation`)

gafter replied:
No.  It is intended that the code gen has some flexibility as to what patterns it recognizes and optimizes.

---
In reply to: [281407592](https://github.com/dotnet/roslyn/pull/35463#discussion_r281407592) [](ancestors = 281407592)

jcouv commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -646,7 +646,15 @@
+    <!-- The type is required an instance of the well-known type System.ReadOnlySpan<T> -->
+    <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
+    <Field Name="Operand" Type="BoundExpression"/>
+    <Field Name="Method" Type="MethodSymbol"/>
```
>Method [](start = 17, length = 6)

nit: this method is the conversion operator used as fallback (when bytes are not in a good shape to be directly pointed at). Maybe add a comment or rename to "ConversionMethod"?

gafter commented:
@jcouv @333fred I've addressed your comments, and added a long description up above (in the PR) of the problem being addressed and the design options for addressing it.  Do you have any further comments?
